otri/downloader/__init__.py

```diff
@@ -808,3 +817,413 @@ def _post_process(self, atom: Mapping, **kwargs) -> Mapping:
+    # TODO: move this validation methods somewhere else
+
+    # DEFAULT PARAMETER VALIDATION METHODS #
+    @staticmethod
+    def match_param_validation(key: str, possible_values: List, required: bool = True) -> Callable:
+        '''
+        Generates a validation method that checks if the parameter's value the possible values for the key.
+        The method raises exception when the parameter's value is NOT between the possible ones.
+
+        Parameters:
+            key : str
+                Only used in the raised exception when value is NOT in the possible ones.
+            possible_values : list
+                Collection of possible values for the parameter.
+            required : bool
+                Whether the parameter is required or not.
+        Returns:
+            A callable validation method.
+        '''
+        def validator(value):
+            if value is None and required:
+                raise ValueError(f"Parameter '{key}' cannot be None")
+            if not required:
+                possible_values.append(None)
+            if value not in possible_values:
+                raise ValueError(f"{value} not a possible value for '{key}', possible values: {possible_values}")
+        return validator
+
+    @staticmethod
+    def datetime_param_validation(key: str, dt_format: str, required: bool = True) -> Callable:
+        '''
+        Generates a validation method that checks if the parameter's value is a datetime with the given format.
+        The method raises exception when the parameter's value is NOT a datetime or in the given format.
+
+        Parameters:
+            key : str
+                Only used in the raised exception when value is wrong.
+            dt_format : str
+                strptime format to parse the parameter's value.
+            required : bool
+                Whether the parameter is required or not.
+        Returns:
+            A callable validation method.
+        '''
+        def validator(value):
+            if value is None and required:
+                raise ValueError(f"Parameter '{key}' cannot be None")
+            if not isinstance(value, str):
+                raise ValueError(f"Parameter '{key}' is not a string")
+            try:
+                datetime.strptime(value, dt_format)
+            except ValueError:
+                raise ValueError(f"Parameter '{key}' with value {value} does not match datetime format {dt_format}")
+        return validator
+
+    # TODO: another default validation could be range check (value in range [min, max])
+    # TODO: another default validation is just checking that a parameter is given
+    # TODO: another default validation is regex.
+
+
+class MappingComp(AdapterComponent):
+    '''
+    Changes parameter's values according to a given mapping.
+    Can be used to reuse the same adapter for different downloads
+    '''
+
+    def __init__(self, key: str, value_mapping: Mapping[str, Set[str]], required: bool = True):
```

Comment:
I may or may not be tripping right now but I think this `value_mapping` could make more sense in reverse, as in mapping values to their translation, so that you can avoid 99% of the work inside `prepare` below.

Something like:
```python
elif self._key in kwargs:
    if kwargs[self._key] not in self._value_mapping:
        raise ValueError(f"Parameter's '{self._key,}' value {kwargs[self._key]} is not in {self._value_mapping}")
    kwargs[self._key] = self._value_mapping[self._key]
```

Considering you take the "first" occurrency in the mapping, which is not known since mappings are not sorted by key, I think having more than one occurrency of the values to translate is not needed and should actually be avoided, hence why having the values to translate as keys makes sense in my head.